django/core/cache/backends/redis.py

```diff
@@ -0,0 +1,239 @@
+                    **self._client_kwargs
+                )
+            else:
+                self._pools[index] = self._pool_class(**params)
+
+        return self._pools[index]
+
+    def get_client(self, key='', write=False):
+        pool = self._get_connection_pool(write)
+        return self._client(connection_pool=pool)
+
+    def get(self, key, default):
+        client = self.get_client(key)
+        value = client.get(key)
+        # If key does not exists, return None
+        if value is None:
+            return default
+        else:
+            return self._serializer.loads(value)
+
+    def add(self, key, value, timeout):
+        client = self.get_client(key)
+        return bool(client.set(key, self._serializer.dumps(value), ex=timeout, nx=True))
+
+    def set(self, key, value, timeout):
+        client = self.get_client(key, write=True)
+        client.set(key, self._serializer.dumps(value), ex=timeout)
+
+    def touch(self, key, timeout):
+        client = self.get_client(key, write=True)
+        if timeout is None:
+            return client.persist(key)
+        else:
+            return client.expire(key, timeout)
+
+    def delete(self, key):
+        client = self.get_client(key, write=True)
+        return bool(client.delete(key))
+
+    def clear(self):
+        client = self.get_client(None, write=True)
+        return client.flushall()
+
+    def get_many(self, keys):
+        client = self.get_client(None)
+        ret = client.mget(keys)
+        return {
+            k: self._serializer.loads(v) for k, v in zip(keys, ret) if v is not None
+        }
+
+    def set_many(self, data, timeout):
+        client = self.get_client(None)
+        serialized_data = {k: self._serializer.dumps(v) for k, v in data.items()}
+
+        client.mset(serialized_data)
+
+        for k, _ in data.items():
+            # Setting timeout for each as redis-py does not support timeout with mset
+            if timeout is None:
+                client.persist(k)
+            else:
+                client.expire(k, timeout)
+
+
+class RedisCache(BaseCache):
+    def __init__(self, server, params):
+        super().__init__(params)
+        if isinstance(server, str):
+            self._servers = re.split('[;,]', server)
+        else:
+            self._servers = server
+
+        self._options = params.get('OPTIONS') or {}
+        self._class = self._options.get('CLIENT', RedisCacheClient)
+
+        if isinstance(self._class, str):
+            self._class = import_string(self._class)
+
+    def get_backend_timeout(self, timeout=DEFAULT_TIMEOUT):
+        # Needs rethinking
+
+        if timeout == DEFAULT_TIMEOUT:
+            timeout = self.default_timeout
+        elif timeout is None:
+            # Redis-py accepts None as timeout for as persistant key
+            return timeout
+        elif timeout <= 0:
+            # Need to fix this
+            timeout = 1
+
+        return int(timeout)
```

Comment:
If `timeout` is `DEFAULT_TIMEOUT` then `self.default_timeout` is what was passed in to the cache as `TIMEOUT`. This could conceivably be `None` so we cannot just use `return int(timeout)`. We should allow returning `None` from here which would allow all keys to be persistent.

Django cache documentation ([1](https://docs.djangoproject.com/en/3.2/topics/cache/#cache-arguments), [2](https://docs.djangoproject.com/en/3.2/ref/settings/#std:setting-CACHES-TIMEOUT)) specifically mentions that the default timeout is 300 seconds. If the value is `None`, then the key will never expire. If the value is `0` then the key will expire immediately.

Looking at the [`EXPIRE`](https://redis.io/commands/expire) documentation for Redis it states:

> Note that calling EXPIRE/PEXPIRE with a non-positive timeout or EXPIREAT/PEXPIREAT with a time in the past will result in the key being deleted rather than expired (accordingly, the emitted key event will be del, not expired).

So it sounds like the default behaviour of Redis is what we would expect, no? And thus we can just do the following:

```suggestion
        if timeout == DEFAULT_TIMEOUT:
            timeout = self.default_timeout

        # The key will be made persistent if None used as a timeout.
        # Non-positive values will cause the key to be deleted.
        return None if timeout is None else int(timeout)
```

I'm aware that Redis supports using `PX` instead of `EX` for millisecond-precision timeouts instead of second-precision. Likewise there is `PEXPIRE` over plain `EXPIRE`. Perhaps for the Redis backend we should allow float values to be provided and always multiply up to milliseconds. This gives more flexibility. I haven't implemented that here, but it is something that could be investigated.

Reply:
`EXPIRE` supports zero and negative timeout but `SET` does not. `SET` only accepts positive non-zero values for expiry time. A work around could be using omitting the use of `ex` and manually using expire to set the timeout.
```
def set(self, key, value, timeout):
        client = self.get_client(key, write=True)
        client.set(key, self._serializer.dumps(value))

        if timeout is None:
            client.persist(key)
        else:
            client.expire(key, timeout)
```
However, this might not be the cleanest approach in my opinion.

Reply:
For the milli-second precision, there was some discussion [here](https://github.com/jazzband/django-redis/pull/508). We can implement both `expire` as well as `pexpire` here too. However, I still feel milli-second precision might not be that useful for the users. Let me know how I shall proceed.

Reply:
> `EXPIRE` supports zero and negative timeout but `SET` does not. `SET` only accepts positive non-zero values for expiry time. A work around could be using omitting the use of `ex` and manually using expire to set the timeout.

I haven't actually tried to see if it works, but the documentation doesn't seem to imply that it does and is as you say. Boo!

Perhaps the way to handle it is to use the atomic operation if we are able, and fall back to separate operations if we cannot:

```python
    def set(self, key, value, timeout):
        client = self.get_client(key, write=True)
        value = self._serializer.dumps(value)
        if timeout is None or timeout > 0:
            client.set(key, value, ex=timeout)
        else:
            # Non-positive timeouts are not supported by SET with EX.
            client.set(key, value)
            client.expire(key, timeout)
```

Note that if `timeout` is `None` we don't need to worry about using `.persist()`. This is because [`SET`](https://redis.io/commands/set) with `ex=None` will discard any existing TTL:

> Any previous time to live associated with the key is discarded on successful SET operation.

This is also mentioned in the documentation for [`EXPIRE`](https://redis.io/commands/expire):

> The timeout will only be cleared by commands that delete or overwrite the contents of the key, including DEL, SET, GETSET and all the *STORE commands.

As long as the value is _replaced_, not merely _altered_ the TTL is removed.

> For the milli-second precision, there was some discussion [here](https://github.com/jazzband/django-redis/pull/508). We can implement both `expire` as well as `pexpire` here too. However, I still feel milli-second precision might not be that useful for the users. Let me know how I shall proceed.

We can leave the millisecond precision stuff for now unless @carltongibson thinks it's worth doing. It'll be an easy follow up as we just need to change `.get_backend_timeout()` to multiply by `1000` and switch over to using `PEXPIRE`, `SET` with `PX`, etc. In other words, we would take `timeout` in fractional seconds and convert to milliseconds for use in the backend internally. (I think it could be useful for users and it has come up in the past - notably memcache doesn't allow for millisecond precision which has been annoying for reducing the duration of certain tests.)

Reply:
Yes this seems to be a better approach. However, for Non-positive timeouts, using `client.set` and then invalidating the key using `client.expire` seems redundant. 
We can keep the function set as
```
def set(self, key, value, timeout):
        client = self.get_client(key, write=True)
        value = self._serializer.dumps(value)
        if timeout is None or timeout > 0:
            client.set(key, value, ex=timeout)
```
This would still produce the same results and reduce the call to the redis-server. Only the case where the given key already exists in the cache, we can just remove the call to `set` and keep the call to `expire` to incorporate this case. Something like this I guess,
```
 def set(self, key, value, timeout):
        client = self.get_client(key, write=True)
        value = self._serializer.dumps(value)
        if timeout is None or timeout > 0:
            client.set(key, value, ex=timeout)
        else:
            # Non-positive timeouts are not supported by SET with EX.
            client.expire(key, timeout)
```
Let me know what do you feel about this?

Reply:
Another modification we would need to do is in the `add` method.
```
    def add(self, key, value, timeout):
        client = self.get_client(key, write=True)
        value = self._serializer.dumps(value)

        if timeout is None or timeout > 0:
            return bool(client.set(key, value, ex=timeout, nx=True))
        else:
            result = bool(client.set(key, value, nx=True))
            if result:
                client.expire(key, timeout)
        return result
```
I think this should do it. as this test fails as if expects True to be returned for a 0 timeout. https://github.com/django/django/blob/8a7ac78b706797a03d26b88eddb9d1067ed35b66/tests/cache/tests.py#L617

Reply:
> Let me know what do you feel about this?

Yes, the `.set()` for non-positive timeouts is pointless. But we still need to expire the key in case it exists. Instead of using `.expire()`, however, we should just go for `.delete()` instead:

```python
    def set(self, key, value, timeout):
        client = self.get_client(key, write=True)
        value = self._serializer.dumps(value)
        if timeout is None or timeout > 0:
            client.set(key, value, ex=timeout)
        else:
            client.delete(key)
```

Using `.expire(key, 0)` would just cause Redis to perform a delete behind the scenes anyway:

> Note that calling EXPIRE/PEXPIRE with a non-positive timeout or EXPIREAT/PEXPIREAT with a time in the past will result in the key being deleted rather than expired (accordingly, the emitted key event will be del, not expired).

Reply:
> Another modification we would need to do is in the `add` method.

Hmm. Yeah, that is not pretty, but I can see the problem. This is also more challenging because it should only set the key if it doesn't exist and we don't want to change the TTL of an existing key. Ideally `ex=0` would be possible to avoid this mess, but I think what you've come up with is the best possible. We can condense it slightly though by using `:=` and use `.delete()` instead of `.expire()`:

```python
    def add(self, key, value, timeout):
        client = self.get_client(key, write=True)
        value = self._serializer.dumps(value)
        if timeout is None or timeout > 0:
            return bool(client.set(key, value, ex=timeout, nx=True))
        else:
            if ret := bool(client.set(key, value, nx=True))
                client.delete(key)
            return ret
```

Instead of a persistent `.set()` followed by a `.delete()`, we could just alter the timeout to use `px=1` so it expires after 1ms? That means the `else` block could be:

```python
            if ret := bool(client.set(key, value, px=1, nx=True))
                client.delete(key)
            return ret
```

Or we could even do away with the `.delete()` as the key will likely have expired before we can execute it:

```python
            return bool(client.set(key, value, px=1, nx=True))
```

Further, if we were to change to using milliseconds, then we can force `0` to be `1` millisecond instead and simplify all of this?

Reply:
```
    return bool(client.set(key, value, px=1, nx=True))
```
Only setting px=1 does not work as this test fails.
https://github.com/django/django/blob/8a7ac78b706797a03d26b88eddb9d1067ed35b66/tests/cache/tests.py#L618
```
 if ret := bool(client.set(key, value, px=1, nx=True))
        client.delete(key)
    return ret
```
The above solution works but setting `px=1` seems redundant and can be removed as we are using `.delete()`